### PR TITLE
[WIP] Feature/multisite

### PR DESCRIPTION
**Pull request by dkarlovi**
Support for multiple apps pushing to the same Xhgui instance, using a separate Mongo collection for each app.

### CODE REVIEW

**Comment by dkarlovi** [comment]
The build seems to fail because of composer.json which I haven't modified. Weird.

File: external/batch-import.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
+<?php
+ini_set('display_errors', 1);
+if (!defined('XHGUI_ROOT_DIR')) {
+    require dirname(__DIR__) . '/src/bootstrap.php';
+}
+
+$sites      = new Xhgui_Sites();
```
Comment (markstory) [inline external/batch-import.php]: What is this aligned with?

File: src/Xhgui/Controller/Run.php [hunk 2]
```diff
@@ -9,6 +9,23 @@ public function __construct($app, $profiles, $watches)
         $this->_watches = $watches;
     }
 
+    public function homepage()
+    {
+        // TODO: how to fetch DIC from app?
+        global $di;
```
Comment (markstory) [inline src/Xhgui/Controller/Run.php]: You don't that defeats the purpose of dependency _injection_. Instead you'd change the controller definitions to inject the new dependency.

File: src/Xhgui/Controller/Run.php [hunk 3]
```diff
@@ -9,6 +9,23 @@ public function __construct($app, $profiles, $watches)
...
+
+        if (false !== ($site = $di['sites']->getFirstAvailable())) {
+            $url = $this->_app->urlFor('home', compact('site'));
+            // TODO: this does not work (?)
+            // $this->_app->redirect($url);
+
+            header('Location: '. $url);
+            exit;
```
Comment (markstory) [inline src/Xhgui/Controller/Run.php]: You probably need to return the result of redirect().

File: src/Xhgui/Sites.php [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    /**
+     * @return array
+     */
+    public function getAvailable()
+    {
+        if (null === $this->_list) {
+            $client = new MongoClient(Xhgui_Config::read('db.host'));
+            $db = $client->selectDB(Xhgui_Config::read('db.db'));
```
Comment (markstory) [inline src/Xhgui/Sites.php]: This seems like a dependency that could be injected.

File: src/Xhgui/Twig/Extension.php [hunk 5]
```diff
@@ -54,6 +57,27 @@ public function truncate($input, $length = 50)
         return substr($input, 0, $length) . "\xe2\x80\xa6";
     }
 
+     public function sites()
+     {
+         // TODO: how to fetch DIC from here?
+         global $di;
```
Comment (markstory) [inline src/Xhgui/Twig/Extension.php]: You don't :smile:

File: src/templates/layout/base.twig [hunk 6]
```diff
@@ -27,6 +27,20 @@
                         <li><a href="{{ url('watch.list') }}">Watch Functions</a></li>
                         <li><a href="{{ url('waterfall.list') }}">Waterfall</a></li>
                     </ul>
+                    {% set sites = sites() %}
```
Comment (markstory) [inline src/templates/layout/base.twig]: This should probably be passed into the template context as part of the template being rendered. That would help solve the issue with globals as well.

File: src/Xhgui/Sites.php [hunk 7]
```diff
@@ -0,0 +1,127 @@
+<?php
+
+/**
+ * Class Xhgui_Sites
+ */
+class Xhgui_Sites
```
Comment (markstory) [inline src/Xhgui/Sites.php]: Any tests we could add here?

Comment (dkarlovi) [inline src/Xhgui/Sites.php]: Sure, this is only to get my feet wet with the project. :)

File: external/batch-import.php (hunk 1 above)
Comment (dkarlovi) [inline external/batch-import.php]: With a snippet of code which I replaced with lines 3-5 (basically the same thing, but I took this from existing source to be more inline, haven't changed the alignment). Will fix it.

File: src/Xhgui/Sites.php (hunk 4 above)
Comment (dkarlovi) [inline src/Xhgui/Sites.php]: True.

File: external/batch-import.php (hunk 1 above)
Comment (markstory) [inline external/batch-import.php]: I find avoiding space alignment entirely saves time and annoyance in the long run :smile:

Comment (dkarlovi) [inline external/batch-import.php]: Totally agree, lately I just setup PHPStom to whatever coding style is used for the projects and keep pressing Ctl-Alt-L. :smile:

File: src/Xhgui/Controller/Run.php (hunk 3 above)
Comment (dkarlovi) [inline src/Xhgui/Controller/Run.php]: Nope, tried that, redirect() return is void, IIRC.

File: src/Xhgui/Controller/Run.php (hunk 2 above)
Comment (dkarlovi) [inline src/Xhgui/Controller/Run.php]: Agree, the "global" part was a quick&dirty solution just to get this to work (I already have my fork deployed in production). Can you point me to a proper way of doing this? I'm not familiar with Slim (only worked with Symfony/Silex and Zend Framework)

File: src/Xhgui/Controller/Run.php (hunk 3 above)
Comment (markstory) [inline src/Xhgui/Controller/Run.php]: But returnning will stop execution here, and slim should handle the redirect.

File: src/Xhgui/Controller/Run.php (hunk 2 above)
Comment (markstory) [inline src/Xhgui/Controller/Run.php]: I've wired all the controllers [into the di container](https://github.com/perftools/xhgui/blob/master/src/Xhgui/ServiceContainer.php#L116-L130). You can add the new dependencies there.

File: src/templates/layout/base.twig (hunk 6 above)
Comment (markstory) [inline src/templates/layout/base.twig]: Another option would be to add the sites into the extension, [when the extensions are added to twig](https://github.com/perftools/xhgui/blob/master/src/Xhgui/ServiceContainer.php#L59).

File: src/Xhgui/Controller/Run.php (hunk 3 above)
Comment (dkarlovi) [inline src/Xhgui/Controller/Run.php]: Problem here is that Twig is trying to render the current page here, from within `redirect()`. Probably trying to render the error page for the exception being thrown. Weird framework, this Slim. :smile:

File: src/templates/layout/base.twig (hunk 6 above)
Comment (dkarlovi) [inline src/templates/layout/base.twig]: You still need to render it in Twig, I would leave the extension as is, having cleaned up the globals from there.

Comment (dkarlovi) [inline src/templates/layout/base.twig]: After removing globals usage, I feel like it's better to do it via Twig functions than via Twig globals or injecting in every template.

File: src/Xhgui/Sites.php (hunk 7 above)
Comment (dkarlovi) [inline src/Xhgui/Sites.php]: Added tests for Twig functions and Xhgui_Sites

File: src/Xhgui/Controller/Run.php (hunk 3 above)
Comment (dkarlovi) [inline src/Xhgui/Controller/Run.php]: Can't figure this one out. I believe this is broken in master too (cannot use `$app->redirect()`)

**Comment by dkarlovi** [comment]
OK, the way I see it, this should be OK to merge. Let me know if you have any more issues with it.

File: src/Xhgui/Controller/Run.php (hunk 3 above)
Comment (markstory) [inline src/Xhgui/Controller/Run.php]: Ok, I can look into that.

**Comment by dkarlovi** [comment]
Hi, are there any problems with this PR that need fixing?

**Comment by markstory** [comment]
@dkarlovi I've not had a chance to look it over again, I'm hoping to get to it in the next few days.

**Comment by dkarlovi** [comment]
Hi, me again. Do you think you could get this merged? I'm looking into adding another feature so dunno where to start the branch from, your master?

**Comment by markstory** [comment]
@dkarlovi I should have some time next week to review this and get it merged in.

**Comment by markstory** [comment]
I finally found some time to check this set of changes out locally. And ran into a few problems:

* The default collector (external/header.php) doesn't store data into site specific collections.
* When there aren't any sites recorded the default homepage creates an infinite redirect.

I couldn't figure out how to get data into the `sites` collection through normal operation of the application, or collector either.

**Comment by dkarlovi** [comment]
OK, will look into it, probably today. Thanks.

**Comment by mallardduck** [comment]
This looks like a really exciting feature, I hope that progress can be continued on this for 1.5!

With this feature added I could see a situation where you can host one instance of xhgui and mongodb on a remote server and then have multiple domains (or even servers) pushing their profile data into the xhprof mongo db table.

That could be really cool! Profiling all your PHP apps, no matter where they are, via a single instance of Xhgui.

**Comment by dkarlovi** [comment]
@mallardduck that's exactly how we're using it. Sites store payloads to a local storage and a cron job runs every five minutes and pushed everything since the last run to a remote XHGUI instance via an "API".

You can add a new site easily while keeping the requirements on the measured application as low as possible.

**Comment by mallardduck** [comment]
That sounds amazing! I'm not sure if I'll have any time to play with this PR myself, but I'd live to see it merged in.

I think that once things calm down at work for me I'll play around with it!

**Comment by markstory** [comment]
I've not had much time to review/merge this obviously. I can try to make some time in the new year as I can see how multi-site support can be really useful.

**Pull request closed without merging**